### ISSUE TITLE
MinGW compiler executables

### ISSUE THREAD

**Issue by siffiejoe** [keplerproject/luarocks#362]
Does anyone remember why we use `mingw32-gcc` instead of `gcc` as default compiler in MinGW? I ask because I have a MinGW installation ([win-builds][1]) that does not have `mingw32-gcc` (available are `cc`, `gcc`, `gcc-4.8.3`, `i686-w64-mingw32-gcc.exe`, and some other unusual names). The pending pull request for static libraries (#264) also adds unprefixed commands (`ar` and `ranlib`). Should we change the default `CC`/`LD` variables for MinGW to plain `gcc`?
Note: `mingw32-make` is still necessary -- no MinGW installation I encountered had a plain `make` executable (I think one is included in MSYS though).

  [1]: http://win-builds.org/

**Comment by Tieske** [comment]
Maybe a test with a fallback?

Op 26 apr. 2015 19:42 schreef Philipp Janda <notifications@github.com>:

Does anyone remember why we use mingw32-gcc instead of gcc as default compiler in MinGW? I ask because I have a MinGW installation (win-builds<http://win-builds.org/>) that does not have mingw32-gcc (available are cc, gcc, gcc-4.8.3, i686-w64-mingw32-gcc.exe, and some other unusual names). The pending pull request for static libraries (#264<https://github.com/keplerproject/luarocks/pull/264>) also adds unprefixed commands (ar and ranlib). Should we change the default CC/LD variables for MinGW to plain gcc?
Note: mingw32-make is still necessary -- no MinGW installation I encountered had a plain make executable (I think one is included in MSYS though).

**Comment by siffiejoe** [comment]
> Maybe a test with a fallback?

During installation or during runtime? Is there anything wrong with plain gcc?

**Comment by stevedonovan** [comment]
It depends who's packaging GCC for Windows.  I use the excellent TDM
distribution and there is just plain gcc.exe available, but also explicitly
named x86_64-w64-mingw32-gcc.exe.  Somewhat confusingly, there is no
make.exe, just  mingw32-make.exe.

**Comment by siffiejoe** [comment]
Ok, so we now know about [TDM][1] (no `mingw32-gcc.exe`), [original MinGW][2] (does have `mingw32-gcc.exe`), and [win-builds][3] (no `mingw32-gcc.exe`). There are others like [nuwen][4], [mingw-builds][5], and [Cygwin][6]. Does anyone have those installed? Are there more?

  [1]: http://tdm-gcc.tdragon.net/
  [2]: http://www.mingw.org/
  [3]: http://win-builds.org/doku.php
  [4]: http://nuwen.net/mingw.html
  [5]: http://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win32/Personal%20Builds/mingw-builds/installer/
  [6]: http://cygwin.com/

**Comment by hishamhm** [comment]
> During installation or during runtime?

If you're going to check, I'd rather add a test at installation time and set the config accordingly.

But if gcc.exe is available everywhere (or mostly everywhere except some obscure configuration -- that's what configurability is for), I'd favor to just switch the default.

**Comment by ignacio** [comment]
nuwen (as of version 12.2) has gcc and make. Can't comment on mingw-builds or Cygwin.

**Comment by Tieske** [comment]
Does plain `gcc.exe
at install time, the binaries are tested for 32/64bit, so maybe we should check at install time?

**Comment by Tieske** [comment]
Does plain `gcc.exe` auto detect 32/64 bit?

At install time, the Lua binaries are tested for 32/64bit, so probably best to check at install time what is available?

This is what I find for MinGW;
````
C:\MinGW\bin>dir *gcc*.exe
 Het volume in station C heeft geen naam.
 Het volumenummer is 809F-7FC6

 Map van C:\MinGW\bin

05-10-2013  19:17            65.550 gcc-ar.exe
05-10-2013  19:17            65.038 gcc-nm.exe
05-10-2013  19:17            65.038 gcc-ranlib.exe
05-10-2013  19:17         1.819.150 gcc.exe
05-10-2013  19:17         1.819.150 mingw32-gcc-4.8.1.exe
05-10-2013  19:17            65.550 mingw32-gcc-ar.exe
05-10-2013  19:17            65.038 mingw32-gcc-nm.exe
05-10-2013  19:17            65.038 mingw32-gcc-ranlib.exe
05-10-2013  19:17         1.819.150 mingw32-gcc.exe
               9 bestand(en)        5.848.702 bytes
               0 map(pen)  10.526.003.200 bytes beschikbaar

C:\MinGW\bin>
````
and for `make`;
````
C:\MinGW\bin>dir *make*.*
 Het volume in station C heeft geen naam.
 Het volumenummer is 809F-7FC6

 Map van C:\MinGW\bin

02-09-2012  02:42           219.662 mingw32-make.exe
               1 bestand(en)          219.662 bytes
               0 map(pen)  10.525.966.336 bytes beschikbaar

C:\MinGW\bin>
````

**Comment by ignacio** [comment]
Is that question for me? I have no idea. How should I test that?

**Comment by Tieske** [comment]
not specifically, but I think it needs an answer to get to a sensible solution

**Comment by siffiejoe** [comment]
> Does plain gcc.exe auto detect 32/64 bit?

I don't think so. How are you supposed to decide when your only input is a C file?!

**Comment by Tieske** [comment]
... and the libraries linked against, which contain the information (but that is linking...)

The installer checks the Lua binaries upon installation.

**Comment by Tieske** [comment]
What executables are needed? is it enough to have `gcc` and `make` or do we also need to go out and look for `ranlib` and others? if so, which ones then?

From the naming mess it seems we need to provide some sort of list (2 lists actually; 32 and 64bit)

so for a workflow;

- test 32/64 bit, based on Lua binaries
- scan system path for `*gcc*.exe` and `*make*.exe` then match them so we know both tools are in the same directory (to prevent mixing tools from different distributions)
- match the results against a know list of distributions
- if last match failed; dump search results so we can quickly update the embedded lists based on reports.

could that work?

**Comment by siffiejoe** [comment]
I think only `gcc` is needed for now (for static linking one also needs `ar` and `ranlib`, but I think that those are -- like `make` -- optional and probably bit-independent).

**Comment by Tieske** [comment]
Here's some code; https://gist.github.com/Tieske/95178193ebb164b7f74f 

It searches the system path for `gcc`, `ar`, `ranlib` and `make` executables located in the same directory.
This is what I get with three suites in my system path;
````
Program starting as '"C:\Program Files (x86)\ZeroBraneStudio\bin\lua.exe" -e "io.stdout:setvbuf('no')" "C:\Users\Thijs\Desktop\filelist.lua"'.
Program 'lua.exe' started in 'C:\Users\Thijs\Dropbox\Lua projects' (pid: 7008).
Complete suites:
path = c:\mingw\bin
	ar=c:\mingw\bin\ar.exe
	gcc=c:\mingw\bin\gcc.exe
	make=c:\mingw\bin\mingw32-make.exe
	ranlib=c:\mingw\bin\ranlib.exe
path = c:\tdm-gcc-32\bin
	ar=c:\tdm-gcc-32\bin\ar.exe
	gcc=c:\tdm-gcc-32\bin\gcc.exe
	make=c:\tdm-gcc-32\bin\mingw32-make.exe
	ranlib=c:\tdm-gcc-32\bin\ranlib.exe
path = c:\tdm-gcc-64\bin
	ar=c:\tdm-gcc-64\bin\ar.exe
	gcc=c:\tdm-gcc-64\bin\gcc.exe
	make=c:\tdm-gcc-64\bin\mingw32-make.exe
	ranlib=c:\tdm-gcc-64\bin\ranlib.exe
Program completed in 3.67 seconds (pid: 7008).
````

This could be integrated in the installer, if the `/MW` switch is specified.
